sys/congure/quic/congure_quic.c

```diff
@@ -0,0 +1,289 @@
+                    if (earliest > ptr->send_time) {
+                        earliest = ptr->send_time;
+                    }
+                    if ((latest - earliest) > pc_duration) {
+                        /* in persistent congestion */
+                        c->super.cwnd = c->consts->min_wnd;
+                        if (c->ssthresh < c->consts->min_wnd) {
+                            /* See
+                             * https://github.com/quicwg/base-drafts/issues/4826#issuecomment-776305871
+                             * XXX: this differs from the pseudo-code in
+                             * Appendix B.8, where when `ssthresh` is lower than
+                             * `cwnd` (e.g. because )
+                             */
+                            c->ssthresh = c->consts->min_wnd;
+                        }
+                        c->recovery_start = 0;
+                    }
+                }
+            } while ((&ptr->super) != msgs->super.next);
+        }
+    }
+}
+
+static void _dec_flight_size(congure_quic_snd_t *c, unsigned msg_size)
+{
+    /* check for integer underflow */
+    if ((c->in_flight_size - msg_size) > c->in_flight_size) {
+        c->in_flight_size = 0U;
+    }
+    else {
+        c->in_flight_size -= msg_size;
+    }
+}
+
+static void _snd_init(congure_snd_t *cong, void *ctx)
+{
+    congure_quic_snd_t *c = (congure_quic_snd_t *)cong;
+
+    c->super.ctx = ctx;
+    c->first_rtt_sample = 0;
+    c->super.cwnd = c->consts->init_wnd;
+    c->in_flight_size = 0U;
+    c->recovery_start = 0U;
+    c->ssthresh = CONGURE_WND_SIZE_MAX;
+    c->limited = 0U;
+    c->max_ack_delay = 0U;
+    c->smoothed_rtt = c->consts->init_rtt;
+    c->rtt_var = c->consts->init_rtt / 2U;
+    c->min_rtt = 0U;
+}
+
+static int32_t _snd_inter_msg_interval(congure_snd_t *cong, unsigned msg_size)
+{
+    congure_quic_snd_t *c = (congure_quic_snd_t *)cong;
+
+    /* interval in QUIC spec is a divisor, so flip denominator and numerator;
+     * smoothed_rtt is in ms, but expected result is in us */
+    return (c->consts->inter_msg_interval_denominator * c->smoothed_rtt *
+            msg_size * US_PER_MS) /
+           (c->consts->inter_msg_interval_numerator * c->super.cwnd);
+}
+
+static void _snd_report_msg_sent(congure_snd_t *cong, unsigned sent_size)
+{
+    congure_quic_snd_t *c = (congure_quic_snd_t *)cong;
+
+    if ((c->in_flight_size + sent_size) <= c->super.cwnd) {
+        c->in_flight_size += sent_size;
+    }
+
+    c->in_flight_size += sent_size;
+}
+
+static void _snd_report_msg_discarded(congure_snd_t *cong, unsigned msg_size)
+{
+    congure_quic_snd_t *c = (congure_quic_snd_t *)cong;
+
+    assert(msg_size <= c->in_flight_size);
+
+    _dec_flight_size(c, msg_size);
+}
+
+static void _snd_report_msgs_lost(congure_snd_t *cong, congure_snd_msg_t *msgs)
+{
+    congure_quic_snd_t *c = (congure_quic_snd_t *)cong;
+    /* XXX need to untangle clist_foreach() to record last_lost_sent */
```

Comment:
What does that mean?

Reply:
These comments are to prevent the reader from thinking "why isn't `clist_foreach()` used here? It basically copyies its code and add stuff".